Fix `is_monotonic` to handle functions with isolated stationary points.

utam-1 opened the pull request:
The old implementation was too strict: it checked for any derivative roots in the interval and immediately returned False. But a function can be monotonic even when its derivative touches zero at isolated points - it just needs to maintain a consistent sign (>= 0 or <= 0) throughout the interval.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #28571 

#### Brief description of what is fixed or changed
The implementation of `is_monotonic` was changed as follows : – 
- Check if derivative >= 0 across the interval (non-decreasing)
- Check if derivative <= 0 across the interval (non-increasing)
- Return True if either condition holds for the entire interval

Test coverage for `is_monotonic` inside sympy\calculus\tests\test_singularities.py was extended to include the following: –
- Added tests for functions with isolated stationary points.
- Added tests for polynomials with derivative zeros.
- Added tests for restricted intervals.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* calculus
  * Fixed `is_monotonic` to correctly identify functions as monotonic when they have isolated stationary points.
<!-- END RELEASE NOTES -->

utam-1 committed:
Fix is_monotonic to handle functions with isolated stationary points

The is_monotonic function was incorrectly rejecting monotonic functions
that have isolated stationary points. For example, x**3 is strictly
increasing everywhere, but returned False simply because f'(0) = 0.

The old implementation was too strict: it checked for any derivative roots
in the interval and immediately returned False. But a function can be
monotonic even when its derivative touches zero at isolated points - it
just needs to maintain a consistent sign (>= 0 or <= 0) throughout the
interval.

This fix replaces the root-finding approach with proper inequality solving:
- Check if derivative >= 0 across the interval (non-decreasing)
- Check if derivative <= 0 across the interval (non-increasing)
- Return True if either condition holds for the entire interval

utam-1 committed:
Removed unicode.

utam-1 committed:
removed sin and cos

bjodah commented:
So... was this Sonnet 4.5 or GPT-5?

utam-1 committed:
Removed comments and test cases.

Removed verbose comments and redundant test cases as suggested by the reviewers.

bjodah commented:
I think using an LLM is fine (I do so too!), but I think it's important that we are all upfront about it when we are doing so (and how). It makes the reviewing process better. Getting back to the contents of this PR in particular: I think the changes here look promising, two things that come to mind are:
1. The new implementation looks quite similar to what's being done in the helper method, except for the interval argument being passed to solveset (I think): is it possible to make use of the existing helper function, or simply delegate to two calls to `is_increasing` and `is_decreasing`? (if not we could perhaps aim for adding a comment in the `is_montonic` source code, explaining why).
2. The new implementation will always call solveset twice, even if the first check would allow us to return early.

Peter230655 commented:
I must admit I fail to see the point of this PR a bit: is it not almost a question of definition whether a function with $\dfrac{df(x)}{dx} \geq 0, x \in (a, b)$ is monotone increasing or strictly monotone increasing in this interval?
But I may be missing the point.

utam-1 committed:
Simplified is_monotonic.

Simplified is_monotonic to delegate checks for increasing or decreasing to the current implementations

utam-1 commented:
> I think using an LLM is fine (I do so too!), but I think it's important that we are all upfront about it when we are doing so (and how). It makes the reviewing process better. Getting back to the contents of this PR in particular: I think the changes here look promising, two things that come to mind are:
> 
> 1. The new implementation looks quite similar to what's being done in the helper method, except for the interval argument being passed to solveset (I think): is it possible to make use of the existing helper function, or simply delegate to two calls to `is_increasing` and `is_decreasing`? (if not we could perhaps aim for adding a comment in the `is_montonic` source code, explaining why).
> 2. The new implementation will always call solveset twice, even if the first check would allow us to return early.

I have made changes on `is_monotonic` as requested, and I appreciate the feedback provided by other reviewers as well regarding the usage of LLMs. I will be careful the next time I submit any PR.
